README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# hlidacr

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/hlidacr)](https://CRAN.R-project.org/package=hlidacr)
[![R build
status](https://github.com/skvrnami/hlidacr/workflows/R-CMD-check/badge.svg)](https://github.com/skvrnami/hlidacr/actions)
[![codecov](https://codecov.io/gh/skvrnami/hlidacr/branch/main/graph/badge.svg?token=FWP73F1DOL)](https://codecov.io/gh/skvrnami/hlidacr)
[![](https://cranlogs.r-pkg.org/badges/hlidacr)](https://CRAN.R-project.org/package=hlidacr)
<!-- badges: end -->

The goal of hlidacr is to provide access to the data published by
[Hlídač státu](https://www.hlidacstatu.cz/) provided by their
[API](https://www.hlidacstatu.cz/api/v2/swagger/index).

## Installation

You can install the package from CRAN:

``` r
install.packages("hlidacr")
```

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("skvrnami/hlidacr")
```

Besides the installation, you need access token for making API requests.
The access token is available
[here](https://www.hlidacstatu.cz/api/v1/Index) after the registration
at the Hlídač státu’s website.

## Usage

The package implements functions for accessing all publicly available
API endpoints as defined in the
[documentation](https://www.hlidacstatu.cz/api/v2/swagger/index) of the
Hlídač státu API.

The data available via the API are related to:

-   funding of political parties  
-   transcriptions of parliamentary floor debates, municipal council
    debates and other bodies  
-   contracts between public institutions and private companies  
-   subsidies  
-   availability of websites operated by public institutions  
    etc.

## Example

There are two types of datasets available at the Hlídač státu API.
First, there are various collection of datasets whose list is available
using `get_datasets()`.

``` r
library(dplyr)
library(hlidacr)

# Authorization token
TOKEN <- Sys.getenv("HLIDAC_TOKEN")

datasets <- get_datasets(token = TOKEN)
str(datasets, max.level = 1)
#> List of 3
#>  $ total  : int 31
#>  $ page   : int 1
#>  $ results:'data.frame': 31 obs. of  16 variables:

head(datasets$Results[,1:2], 10)
#> NULL
```

In general, the data are usually returned in a list with 3 elements:

-   `page`  
-   `total`: total number of records  
-   `results`: data

(Therefore, to get all the data you need to iterate over the pages by
specifying particular page of the results you want)

The data from these datasets can be obtained by `get_dataset_data`
function using dataset’s ID. For example:

``` r
ministers <- get_dataset_data("ministri", page = 1)
head(ministers$results %>% select(resort, jmeno, strana, zacatek))
#>                                 resort            jmeno   strana
#> 1                  Ministerstvo obrany Miloslav Výborný  KDU-ČSL
#> 2     Ministerstvo životního prostředí   Richard Brabec ANO 2011
#> 3       Ministerstvo zahraničních věcí        Jan Kavan     ČSSD
#> 4           Ministerstvo spravedlnosti     Pavel Blažek      ODS
#> 5                       Předseda vlády    Jiří Paroubek     ČSSD
#> 6 Ministerstvo práce a sociálních věcí  Zdeněk Škromach     ČSSD
#>                     zacatek
#> 1 1996-07-04T00:00:00+02:00
#> 2 2018-06-27T00:00:00+02:00
#> 3 1998-07-22T00:00:00+02:00
#> 4 2012-07-03T00:00:00+02:00
#> 5 2005-04-25T00:00:00+02:00
#> 6 2004-08-04T00:00:00+02:00
```

Second, there are specific datasets that are available via specific
routes and therefore specific functions. These include datasets related
to contracts between public institutions and private companies,
subsidies, companies, persons and websites.

For example:

``` r
golf_subsidies <- search_subsidies("golf")
head(golf_subsidies$results %>% select(idDotace, nazevProjektu, dotaceCelkem))
#>                                       idDotace
#> 1 dotinfo-2ebb2071-8444-40be-8560-91af6e7fefa0
#> 2                         deminimis-1000229862
#> 3                         deminimis-1000411828
#> 4       eufondy-04-06-cz-04-1-05-4-1-44-2-2970
#> 5                         deminimis-1000219471
#> 6                         deminimis-1000359004
#>                                                nazevProjektu dotaceCelkem
#> 1    GOLF CLUB U Hrádečku - podnikové vzdělávání zaměstnanců            0
#> 2               Mezinárodní golfový turnaj v extrémním golfu         5000
#> 3                                           Golf pro všechny       200000
#> 4                                           Golf pro všechny       360555
#> 5                                             Golfový turnaj        30000
#> 6 Mezinárodní golfový turnaj 112 ANNIVERSARY GOLF TOURNAMENT       150000
```

``` r
golf_contracts <- search_contracts("golf")
head(golf_contracts$results %>% select(predmet, hodnotaBezDph))
#>                                                   predmet hodnotaBezDph
#> 1           836/OSRM/2017 smlouva o dílo - adventure golf        245000
#> 2  Smlouva o poskytování reklamních a propagačních služeb         70000
#> 3          Smlouva o dílo - výroba propagačního materiálu        105000
#> 4  Smlouva o poskytování reklamních a propagačních služeb        762000
#> 5  Smlouva o poskytování reklamních a propagačních služeb        762000
#> 6 Smlouva o poskytování reklamních a propagačních služeb         210000
```

In addition, you can get the text of particular contract using
`get_contract_text` using the ID of the contract that is stored in the
column `id` in the data.frame returned by `search_contracts`.

``` r
con_text <- get_contract_text(id = "9934567")
cat(substr(con_text[1], 1500, 2000))
#>  Smlouvy
#> 
#> 1.1. Prodávající se zavazuje, že Kupujícímu odevzdá za podmínek stanovených touto
#> Smlouvou golfový multisimulátor Visual Sports s golfovým trenažérem s poslední verzí
#> golfu E6 a herní konzolí s nabídkou her jako je fotbal, basketbal, hokej a další možnosti
#> pro využití spokojenosti dětí a laserovou střelnicí, vč. sportovního vybavení na fotbal,
#> hokej a basketbal, které jsou co do technických parametrů a svého příslušenství blíže
#> specifikovány v příloze č. 1 této Smlouvy (dále jen
```

Searching for a person is done using `search_person`. For instance:

``` r
babis <- search_person("Babiš")
head(babis)
#>   titulPred    jmeno prijmeni titulPo            narozeni           nameId
#> 1      <NA>   Patrik    Babiš    <NA> 1973-05-01T00:00:00     patrik-babis
#> 2              Pavel    Babiš         1975-02-20T00:00:00      pavel-babis
#> 3              Miloš    Babiš         1960-05-12T00:00:00      milos-babis
#> 4      <NA>   Patrik    Babiš    <NA> 1972-04-17T00:00:00   patrik-babis-1
#> 5      Ing.   Andrej    Babiš    <NA> 1954-09-02T00:00:00     andrej-babis
#> 6           Alexandr    Babic         1975-11-07T00:00:00 alexandr-babic-2
#>                   profile
#> 1     /osoba/patrik-babis
#> 2      /osoba/pavel-babis
#> 3      /osoba/milos-babis
#> 4   /osoba/patrik-babis-1
#> 5     /osoba/andrej-babis
#> 6 /osoba/alexandr-babic-2
```

`get_person` function provides data related to a person such as their
donations to political parties, service in public and private
institutions and their social media accounts. The example below shows
Andrej Babiš’s donations to political parties.

``` r
ab <- get_person("andrej-babis")
head(ab$sponzoring %>% arrange(desc(castka)))
#>       typ organizace role  castka             datumOd             datumDo
#> 1 sponzor        ANO   NA 8000000 2012-01-01T00:00:00 2012-01-01T00:00:00
#> 2 sponzor        ANO   NA 2500000 2012-01-01T00:00:00 2012-01-01T00:00:00
#> 3 sponzor        ANO   NA 2500000 2012-01-01T00:00:00 2012-01-01T00:00:00
#> 4 sponzor        ANO   NA 2000000 2012-01-01T00:00:00 2012-01-01T00:00:00
#> 5 sponzor        ANO   NA 2000000 2012-01-01T00:00:00 2012-01-01T00:00:00
#> 6 sponzor        ANO   NA 2000000 2012-01-01T00:00:00 2012-01-01T00:00:00
```

Besides getting social media accounts for a particular person using the
`get_person` function, you can obtain social media accounts of Czech
politicians, using `get_person_social`. It returns a data.frame with a
data.frame nested in the `socialniSite` variable. For instance:

``` r
twitter_insta <- get_person_social(types = c("Instagram", "Twitter"))
twitter_insta$socialniSite[2]
#> [[1]]
#>        type           id                                    url
#> 1   Twitter PatrikNacher       https://twitter.com/PatrikNacher
#> 2 Instagram nacherpatrik https://www.instagram.com/nacherpatrik
```


NEWS.md
# NEWS

## hlidacr 0.2.0
- make error message more informative
- update documentation 
    - inform users about `page` limits
    - update for changes in API
    - update README
- add `search_person_by_parameters` function

## hlidacr 0.1.3
- updates based on CRAN submission
    - update DESCRIPTION
    - update documentation

## hlidacr 0.1.2
- add User Agent to headers
- use query in httr GETs in search functions

## hlidacr 0.1.1

- use environment variable `HLIDAC_TOKEN` by default as token in all functions
- update README to include more examples


tests/testthat.R
library(testthat)
library(hlidacr)

test_check("hlidacr")


tests/testthat/test-api.R
test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})

response_401 <- list(status_code = 401)
response_404 <- list(status_code = 404)

test_that("error is handled properly", {
    expect_error(handle_error_response(response_401),
                   message = "HTTP Error 401 when connecting to Hlidac statu. Probably invalid token.")
    expect_error(handle_error_response(response_404),
                   message = "HTTP Error 404 when connecting to Hlidac statu. Not found.")
})


test_that("queries are created properly", {
    expect_equal(create_query("page=1"), "page=1")
    expect_equal(create_query("page=1", "sort=0"), "page=1&sort=0")
})


tests/testthat/test-vcr.R
test_that("get_datasets() works", {
    vcr::use_cassette("get_datasets", {
        datasets <- get_datasets(token = Sys.getenv("HLIDAC_TOKEN"))
    })
    testthat::expect_type(datasets, "list")
    testthat::expect_type(datasets$Total, "integer")
    testthat::expect_true(datasets$Total > 0)
    testthat::expect_type(datasets$Page, "integer")
    testthat::expect_equal(datasets$Total, nrow(datasets$Results))
})

test_that("get_dataset_metadata() works", {
    vcr::use_cassette("get_dataset_metadata", {
        dataset_metadata <- get_dataset_metadata("ministri",
                                                 token = Sys.getenv("HLIDAC_TOKEN"))
    })
    testthat::expect_type(dataset_metadata, "list")
    testthat::expect_type(dataset_metadata$id, "character")
})

test_that("get_dataset_data() works", {
    vcr::use_cassette("get_dataset_data", {
        dataset_data <- get_dataset_data("ministri",
                                         token = Sys.getenv("HLIDAC_TOKEN"))
        dataset_query <- get_dataset_data("ministri",
                                          token = Sys.getenv("HLIDAC_TOKEN"),
                                          query = "Zeman")
    })

    testthat::expect_type(dataset_data, "list")
    testthat::expect_type(dataset_data$Total, "integer")
    testthat::expect_true(dataset_data$Total > 0)
    testthat::expect_type(dataset_data$Page, "integer")
    testthat::expect_gt(nrow(dataset_data$Results), 0)

    testthat::expect_type(dataset_query, "list")
    testthat::expect_type(dataset_query$Total, "integer")
    testthat::expect_true(dataset_query$Total > 0)
    testthat::expect_type(dataset_query$Page, "integer")
    testthat::expect_gt(nrow(dataset_query$Results), 0)

})

test_that("get_dataset_record_detail() works", {
    vcr::use_cassette("get_dataset_record_detail", {
        rec_detail <- get_dataset_record_detail("ministri", 1,
                                                token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(rec_detail, "list")
    testthat::expect_type(rec_detail$vlada, "character")
})

test_that("search_subsidies() works", {
    vcr::use_cassette("search_subsidies", {
        golf_sub <- search_subsidies("golf", page = 1,
                                       token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(golf_sub, "list")
    testthat::expect_equal(nrow(golf_sub$Results), 25)
})

test_that("get_subsidy() works", {
    vcr::use_cassette("get_subsidy", {
        golf_sub1 <- get_subsidy("deminimis-1000229862",
                                 token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(golf_sub1, "list")
    testthat::expect_type(golf_sub1$IdDotace, "character")
    testthat::expect_type(golf_sub1$Prijemce, "list")
})

test_that("get_company() works", {
    vcr::use_cassette("get_company", {
        company <- get_company("Agrofert", token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(company, "list")
    testthat::expect_type(company$ICO, "character")
    testthat::expect_type(company$Jmeno, "character")
})

test_that("get_person() works", {
    vcr::use_cassette("get_person", {
        person <- get_person("andrej-babis", token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(person, "list")
    testthat::expect_type(person$Jmeno, "character")
    testthat::expect_type(person$Prijmeni, "character")
    testthat::expect_gt(nrow(person$SocialniSite), 0)
    testthat::expect_gt(nrow(person$Sponzoring), 0)
    testthat::expect_gt(nrow(person$Udalosti), 0)
})

test_that("search_person() works", {
    vcr::use_cassette("search_person", {
        found_person1 <- search_person(query = "Kalousek",
                                       party = "TOP 09",
                                       token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_gt(nrow(found_person1), 0)
})

test_that("get_person_social() works", {
    vcr::use_cassette("get_person_social", {
        socials <- get_person_social(c("Twitter", "Instagram"),
                                     token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_gt(nrow(socials), 0)
})

test_that("search_contracts() works", {
    vcr::use_cassette("search_contracts", {
        contracts <- search_contracts("golf",
                                      token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(contracts, "list")
    testthat::expect_type(contracts$Total, "integer")
    testthat::expect_type(contracts$Page, "integer")
    testthat::expect_gt(nrow(contracts$Results), 0)
})

test_that("get_contract() works", {
    vcr::use_cassette("get_contract", {
        contract1 <- get_contract("3583464",
                                  token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(contract1, "list")
})

test_that("get_contract_text() works", {
    vcr::use_cassette("get_contract_text", {
        contract_text <- get_contract_text("3583464",
                                           token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(contract_text, "character")
})

test_that("get_websites() works", {
    vcr::use_cassette("get_websites", {
        webs <- get_websites(token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_gt(nrow(webs), 0)
})

test_that("get_website_detail() works", {
    vcr::use_cassette("get_website_detail", {
        web <- get_website_detail(id = "10107",
                                   token = Sys.getenv("HLIDAC_TOKEN"))
    })

    testthat::expect_type(web, "list")
    testthat::expect_type(web$SSL, "list")
    testthat::expect_gt(nrow(web$Availability$Data), 0)
})


R/api.R
#' Get list of datasets
#'
#' Get list of datasets available via 'Hlídač státu' API
#'
#' @param token Authorization token
#'
#' @return Response from the server containing a list of all datasets available.
#' The list contains 3 elements:
#' - total: Total number of datasets available (integer)
#' - page: Page of the result (integer)
#' - results: data concerning the datasets (data.frame)
#' @export
#' @family Datasets
#' @examples
#' \dontrun{
#' get_datasets()
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_datasets <- function(token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()
  response <- httr::GET(
    "https://www.hlidacstatu.cz/Api/v2/datasety/",
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get dataset's metadata
#'
#' Get metadata related to specified dataset.
#'
#' @param dataset_id ID of dataset
#' @param token Authorization token
#'
#' @return a list containing metadata related to the specified dataset
#' @export
#' @family Datasets
#' @examples
#' \dontrun{
#' get_dataset_metadata("ministri")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_dataset_metadata <- function(dataset_id, token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()
  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/Api/v2/datasety/{dataset_id}"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get data from the dataset
#'
#' @param dataset_id ID of dataset
#' @param token Authorization token
#' @param query Search query
#' @param page Number of page (Please note that if the page parameter exceeds 200, the function returns error)
#' @param sort Sort by column
#' @param desc Descending or ascending sorting
#'
#' @return list containing 3 elements:
#' - total: Total number of datasets available (integer)
#' - page: Page of the result (integer), equal to the `page` argument of the function
#' - results: Data.frame with data, columns vary depending on the dataset
#' @export
#' @family Datasets
#' @examples
#' \dontrun{
#' get_dataset_data("ministri")
#' get_dataset_data("ministri", query = "Zeman", page = 1)
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_dataset_data <- function(dataset_id, token = Sys.getenv("HLIDAC_TOKEN"),
                             query = NULL, page = 1, sort = NULL, desc = NULL) {
  check_token(token)
  check_connection()

  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/Api/v2/datasety/{dataset_id}/hledat"),
    query = list(dotaz = query,
                 strana = page,
                 sort = sort,
                 desc = desc),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get record detail
#'
#' @param dataset_id ID of dataset
#' @param item_id ID of record
#' @param token Authorization token
#'
#' @return list of vectors containing data related to the item
#' @export
#' @family Datasets
#' @examples
#' \dontrun{
#' get_dataset_record_detail("ministri", item_id = 1)
#' }
#' @seealso
#' \code{\link{get_dataset_data}}
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_dataset_record_detail <- function(dataset_id, item_id,
                                      token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()

  response <- httr::GET(
    glue::glue(
      "https://www.hlidacstatu.cz/Api/v2/datasety/{dataset_id}/zaznamy/{item_id}"
    ),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Search subsidies
#'
#' @param query Search query
#' @param page Page number (Please note that if the page parameter exceeds 200, the function returns error)
#' @param sort Sorting of results, the available options are the following:
#' - 0: order by relevance
#' - 1: order by the date of signature, the most recent first
#' - 2: order by the date of signature, the most recent last
#' - 3: order by the size of the subsidy, the largest first
#' - 4: order by the size of the subsidy, the largest last
#' - 5: order by ICO in a descending order
#' - 6: order by ICO in an ascending order
#' @param token Authorization token
#'
#' @return list containing 3 elements:
#' - total: Total number of datasets available (integer)
#' - page: Page of the result (integer), equal to the `page` argument of the function
#' - results: Data.frame with data, columns vary depending on the dataset
#' @export
#' @family Subsidies
#' @examples
#' \dontrun{
#' search_subsidies("golf")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
search_subsidies <- function(query, page = 1, sort = NULL,
                             token = Sys.getenv("HLIDAC_TOKEN")){
  check_token(token)
  check_connection()

  response <- httr::GET(
    "https://www.hlidacstatu.cz/api/v2/dotace/hledat",
    query = list(dotaz = query,
                 strana = page,
                 razeni = sort),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get subsidy
#'
#' @param id Subsidy ID
#' @param token Authorization token
#'
#' @return list with details of the subsidy
#' @export
#' @family Subsidies
#' @examples
#' \dontrun{
#' get_subsidy("deminimis-1000229862")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_subsidy <- function(id, token = Sys.getenv("HLIDAC_TOKEN")){
  check_token(token)
  check_connection()

  response <- httr::GET(
    glue::glue(
      "https://www.hlidacstatu.cz/api/v2/dotace/{id}"
    ),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get company details
#'
#' @param company_name Name of the company
#' @param token Access token
#' @return list with 3 elements:
#' - ico: ID number of the company
#' - jmeno: Name of the company
#' - datoveSchranky: Data post box - electronic repository for communication between
#' public institutions and private companies/people
#' @export
#' @family Companies
#' @examples
#' \dontrun{
#' get_company("Agrofert")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_company <- function(company_name, token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()

  company_name <- urltools::url_encode(company_name)
  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/api/v2/firmy/{company_name}"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get person details
#'
#' @param person_id Person's ID
#' @param token Access token
#' @return list with the following elements:
#' - titulPred: academic titles written before person's name
#' - jmeno: given name
#' - prijmeni: surname
#' - titulPo: academic titles written after person's name (if applicable)
#' - narozeni: date of birth
#' - nameId: person's id
#' - profile: URL of profile at hlidacstatu.cz
#' - sponzoring: data.frame with data on sponsorship of political parties
#' by the person
#' - udalosti: events related to the person such as party membership and
#' running in election, serving in political bodies and relations to private
#' companies (shareholder, serving on a board)
#' - socialniSite: data.frame with data on social media accounts
#' @export
#' @family Persons
#' @examples
#' \dontrun{
#' get_person("andrej-babis")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_person <- function(person_id, token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()

  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/api/v2/osoby/{person_id}"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Search person
#'
#' @param query Search query for a person, required parameter
#' @param page Page number
#' @param token Access token
#'
#' @return data.frame with found persons with the following columns:
#' - jmeno: given name
#' - prijmeni: surname
#' - narozeni: date of birth
#' - nameId: person's ID
#' - profile: URL of profile at hlidacstatu.cz
#' - titulPred: academic titles written before person's name
#' - titulPo: academic titles written after person's name
#' @export
#' @family Persons
#' @examples
#' \dontrun{
#' search_person(query = "Kalousek")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
search_person <- function(query, page = NULL,
                          token = Sys.getenv("HLIDAC_TOKEN")){
  check_token(token)
  check_connection()

  if (is.null(query)) {
    usethis::ui_stop("Query is required.")
  }

  response <- httr::GET(
    "https://www.hlidacstatu.cz/api/v2/osoby/hledatFtx",
    query = list(ftxDotaz = query,
                 strana = page),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}


#' Search person by name and date of birth
#'
#' @param first_name Person's first name
#' @param last_name Person's last name
#' @param birth_date Person's date of birth (in YYYY-MM-DD format)
#' @param ignore_diacritics Parameter indicating whether diacritics in names should be ignored (boolean)
#' @param token Access token
#'
#' @return data.frame with found persons with the following columns:
#' - jmeno: given name
#' - prijmeni: surname
#' - narozeni: date of birth
#' - nameId: person's ID
#' - profile: URL of profile at hlidacstatu.cz
#' - titulPred: academic titles written before person's name
#' - titulPo: academic titles written after person's name
#' @export
#' @family Persons
#' @examples
#' \dontrun{
#' search_person_by_parameters(first_name = "Miroslav", last_name = "Kalousek", birth_date = "1960-12-17")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
search_person_by_parameters <- function(first_name = NULL,
                                        last_name = NULL,
                                        birth_date = NULL,
                                        ignore_diacritics = TRUE,
                                        token = Sys.getenv("HLIDAC_TOKEN")){
  check_token(token)
  check_connection()

  if (is.null(first_name)) {
    usethis::ui_stop("first_name is required.")
  }

  if (is.null(last_name)) {
    usethis::ui_stop("last_name is required.")
  }

  if (is.null(birth_date)) {
    usethis::ui_stop("birth_date is required.")
  }

  response <- httr::GET(
    "https://www.hlidacstatu.cz/api/v2/osoby/hledat",
    query = list(jmeno = first_name,
                 prijmeni = last_name,
                 datumNarozeni = birth_date,
                 ignoreDiakritiku = ignore_diacritics),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get person social media accounts
#'
#' @param types Types of social media accounts, only the following types are allowed:
#' 'Twitter', 'Facebook_page', 'Facebook_profile', 'Instagram', 'WWW', 'Youtube'
#' @param token Access token
#' @return data.frame with data on persons' social accounts with the
#' following columns:
#' - titulPred: academic titles written before person's name
#' - jmeno: given name
#' - prijmeni: surname
#' - titulPo: academic titles written after person's name
#' - nameId: person's ID
#' - profile: URL of profile at hlidacstatu.cz
#' - socialniSite: data.frame with data on social accounts (Type of social media,
#' Id of the social media account, URL)
#' @export
#' @family Persons
#' @examples
#' \dontrun{
#' get_person_social(types = c("Twitter", "Instagram"))
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_person_social <- function(types, token = Sys.getenv("HLIDAC_TOKEN")){
  check_token(token)
  check_connection()

  check_types(types)
  opts <- create_type(types)

  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/api/v2/osoby/social?{opts}"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Search contract
#'
#' @param query Full-text search query
#' @param page Page of results (Please note that if the page parameter exceeds 200, the function returns error)
#' @param sort Ordering of the results, the available options are the following:
#' - 0: sort by relevance
#' - 1: recently added contracts first
#' - 2: recently added contracts last
#' - 3: the cheapest contracts first
#' - 4: the most expensive contracts first
#' - 5: recently concluded contracts first
#' - 6: recently concluded contracts last
#' - 7: the most defective contracts first
#' - 8: sort by purchaser
#' - 9: sort by supplier
#' @param token Access token
#'
#' @return list containing 3 elements:
#' - Total: Total number of datasets available (integer)
#' - Page: Page of the result (integer)
#' - Results: Data.frame with data concerning the datasets
#'
#' @export
#' @family Contracts
#' @examples
#' \dontrun{
#' search_contracts(query = "golf")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
#'
#' \url{https://www.hlidacstatu.cz/napoveda}
#'
#' \url{https://smlouvy.gov.cz/}
search_contracts <- function(query, token = Sys.getenv("HLIDAC_TOKEN"),
                             page = 1, sort = 0) {
  check_token(token)
  check_connection()

  if (is.null(query)) {
    usethis::ui_stop("Query is required.")
  }

  response <- httr::GET(
    "https://www.hlidacstatu.cz/api/v2/smlouvy/hledat",
    query = list(dotaz = query,
                 strana = page,
                 razeni = sort),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get contract details
#'
#' @param id ID of contract
#' @param token Authorization token
#'
#' @return list with data
#'
#' @export
#' @examples
#' \dontrun{
#' get_contract(id = "1086905")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
#'
#' \url{https://smlouvy.gov.cz/}
get_contract <- function(id, token = NULL) {
  check_token(token)
  check_connection()
  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/api/v2/smlouvy/{id}"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get contract text
#'
#' @param id ID of contract
#' @param token Authorization token
#'
#' @return Character vector containing text of the contract
#'
#' @export
#' @examples
#' \dontrun{
#' get_contract_text(id = "1086905")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
#'
#' \url{https://smlouvy.gov.cz/}
get_contract_text <- function(id, token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()
  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/api/v2/smlouvy/text/{id}"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get websites
#'
#' @param token Authorization token
#'
#' @return data.frame with data on websites run by the state with the following
#' columns:
#' - hostid: ID of website
#' - host: host of website
#' - url: URL of website
#' - opendataUrl: URL to open data on website's availability
#' - pageUrl: URL to website's profile at Hlidac statu
#' - urad: the institution running the website
#' - publicname: name of the website
#' - popis: description of the website
#' - hash
#'
#' @export
#' @examples
#' \dontrun{
#' get_websites()
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_websites <- function(token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()
  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/api/v2/Weby"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}

#' Get website detail
#'
#' @param id Website id
#' @param token Authorization token
#'
#' @return list with 2 elements:
#' - availability: data on availability of the website
#' including response time and HTTP status code every minute
#' - ssl: assessment of the website's SSL server configuration
#'
#' @export
#' @examples
#' \dontrun{
#' get_website_detail(id = "10107")
#' }
#' @seealso
#' \url{https://www.hlidacstatu.cz/api/v2/swagger/index}
#'
#' \url{https://www.hlidacstatu.cz/api/v1/doc}
get_website_detail <- function(id, token = Sys.getenv("HLIDAC_TOKEN")) {
  check_token(token)
  check_connection()
  response <- httr::GET(
    glue::glue("https://www.hlidacstatu.cz/api/v2/Weby/{id}"),
    httr::add_headers(Authorization = token,
                      user_agent)
  )

  if (response$status_code != 200) {
    handle_error_response(response)
  }

  jsonlite::fromJSON(stringr::str_conv(response$content, "UTF-8"))
}
